mne/preprocessing/artifact_detection.py

```diff
@@ -0,0 +1,220 @@
+        annot += _annotations_from_mask(hp_ts, bad_mask, 'BAD_mov_disp')
+    return annot, disp
+
+
+def compute_average_dev_head_t(raw, pos):
+    """Get new device to head transform based on averaged recording.
+
+    It excludes pos that have a BAD_ annotation.
+    """
+    import warnings
+
+    sfreq = raw.info['sfreq']
+    seg_good = np.ones(len(raw.times))
+    trans_pos = np.zeros(3)
+    hp = pos.copy()
+    hp_ts = hp[:, 0] - raw._first_time
+
+    # Check rounding issues at 0 time
+    if hp_ts[0] < 0:
+        hp_ts[0] = 0
+        assert hp_ts[1] > 1. / sfreq
+
+    # Mask out segments if beyond scan time
+    mask = hp_ts <= raw.times[-1]
+    if not mask.all():
+        warnings.warn(
+            '          Removing %0.1f%% time points > raw.times[-1] (%s)'
+            % ((~mask).sum() / float(len(mask)), raw.times[-1]))
+        hp = hp[mask]
+    del mask, hp_ts
+
+    # Get time indices
+    ts = np.concatenate((hp[:, 0], [(raw.last_samp + 1) / sfreq]))
+    assert (np.diff(ts) > 0).all()
+    ts -= raw.first_samp / sfreq
+    idx = raw.time_as_index(ts, use_rounding=True)
+    del ts
+    if idx[0] == -1:  # annoying rounding errors
+        idx[0] = 0
+        assert idx[1] > 0
+    assert (idx >= 0).all()
+    assert idx[-1] == len(seg_good)
+    assert (np.diff(idx) > 0).all()
+
+    # Mark times bad that are bad according to annotations
+    onsets, ends = _annotations_starts_stops(raw, 'bad')
+    for onset, end in zip(onsets, ends):
+        seg_good[onset:end] = 0
+    dt = np.diff(np.cumsum(np.concatenate([[0], seg_good]))[idx])
+    assert (dt >= 0).all()
+    dt = dt / sfreq
+    del seg_good, idx
+
+    # Get weighted head pos trans and rot
+    trans_pos += np.dot(dt, hp[:, 4:7])
+    rot_qs = hp[:, 1:4]
+    res = 1 - np.sum(rot_qs * rot_qs, axis=-1, keepdims=True)
+    assert (res >= 0).all()
+    rot_qs = np.concatenate((rot_qs, np.sqrt(res)), axis=-1)
+    assert np.allclose(np.linalg.norm(rot_qs, axis=1), 1)
+    rot_qs *= dt[:, np.newaxis]
+    # rank 1 update method
+    # https://arc.aiaa.org/doi/abs/10.2514/1.28949?journalCode=jgcd
+    # https://github.com/tolgabirdal/averaging_quaternions/blob/master/wavg_quaternion_markley.m  # noqa: E501
+    # qs.append(rot_qs)
+    outers = np.einsum('ij,ik->ijk', rot_qs, rot_qs)
+    A = outers.sum(axis=0)
+    dt_sum = dt.sum()
+    assert dt_sum >= 0
+    norm = dt_sum
+    if norm <= 0:
+        raise RuntimeError('No good segments found (norm=%s)' % (norm,))
+    A /= norm
+
+    best_q = linalg.eigh(A)[1][:, -1]  # largest eigenvector is the wavg
+    # Same as the largest eigenvector from the concatenation of all
+    # best_q = linalg.svd(np.concatenate(qs).T)[0][:, 0]
+    best_q = best_q[:3] * np.sign(best_q[-1])
+    trans = np.eye(4)
+    trans[:3, :3] = quat_to_rot(best_q)
+    trans[:3, 3] = trans_pos / norm
+    assert np.linalg.norm(trans[:3, 3]) < 1  # less than 1 meter is sane
+    dev_head_t = Transform('meg', 'head', trans)
```

Comment:
why with  `_average_quats` I get the negative of the eig decomposition? 

   ```
        best_q1 = _average_quats(rot_qs, weights=dt)
   
        rot_qs *= dt[:, np.newaxis]
        # rank 1 update method
        # https://arc.aiaa.org/doi/abs/10.2514/1.28949?journalCode=jgcd
        # https://github.com/tolgabirdal/averaging_quaternions/blob/master/wavg_quaternion_markley.m  # noqa: E501
        # qs.append(rot_qs)
        outers = np.einsum('ij,ik->ijk', rot_qs, rot_qs)
        A = outers.sum(axis=0)
        dt_sum = dt.sum()
        assert dt_sum >= 0
        norm = dt_sum
        if norm <= 0:
            raise RuntimeError('No good segments found (norm=%s)' % (norm,))
        A /= norm
        best_q = linalg.eigh(A)[1][:, -1]  # largest eigenvector is the wavg
        # Same as the largest eigenvector from the concatenation of all
        # best_q = linalg.svd(np.concatenate(qs).T)[0][:, 0]
        best_q = best_q[:3] * np.sign(best_q[-1]) # Proabably here the sing 
```
Then when converting from obscure quaternians to rotation the values for `best_q1` and `best_q` are different.

Reply:
By a sign flip, or something else? There is a sign ambiguity in quaternions. By convention we (usually) take q0 to always be positive, hence the multiplication by the sign at the end.

There is also a bug in average quats where we append the residual element instead of prepend. In the end it doesn't matter because we also take the first three as output (chopping off what we appended) and the calculation ends up being equivalent, but we should fix it. I'll open a quick PR.

Reply:
@AdoNunes you can just use the MNE way of doing things always. If you see it's not equivalent to the code I sent you then there might be some bug in one implementation or the other. You should just use the MNE one (I vetted it more in #7158 so it should be okay), no need to keep the other copy around.

Reply:
still to do?

Reply:
I am just using the mne function  `_average_quats` with the q0 sign convention.